### PR TITLE
Replace response error with choice

### PR DESCRIPTION
**Pull request by ghostbuster91**

### CODE REVIEW

File: json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala [hunk 1]
```diff
@@ -20,8 +22,33 @@ trait SttpCirceApi {
     * - `Left(HttpError(String))` if the response code was other than 2xx (deserialization is not attempted)
     * - `Left(DeserializationError)` if there's an error during deserialization
     */
-  def asJson[B: Decoder: IsOption]: ResponseAs[Either[ResponseError[io.circe.Error], B], Nothing] =
-    asString.mapWithMetadata(ResponseAs.deserializeRightWithError(deserializeJson))
+  def asJson[B: Decoder: IsOption]
+      : ResponseAs[Choice[HttpError[String], DeserializationError[io.circe.Error], B], Nothing] =
+    ???
+
+  def unsafeHttp[L <: Exception, M, R]: Choice[L, M, R] => Choice[Nothing, M, R] = {
```
Comment (ghostbuster91) [inline json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala:29]: Those methods could be exposed on `ResponseAs` via implicit class mechanism

File: json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala [hunk 2]
```diff
@@ -20,8 +22,33 @@ trait SttpCirceApi {
...
+    case Choice.Left(value)   => throw value
+    case Choice.Middle(value) => Choice.Middle(value)
+    case Choice.Right(value)  => Choice.Right(value)
+  }
+
+  def unsafeDeserialization[L, M <: Exception, R]: Choice[L, M, R] => Choice[L, Nothing, R] = {
```
Comment (ghostbuster91) [inline json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala:35]: Same here

File: json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala [hunk 3]
```diff
@@ -20,8 +22,33 @@ trait SttpCirceApi {
...
+  {
+    val l: Either[Nothing, String] = ???
+    l.merge
+    import sttp.client.circe.SttpCirceApi._
+
+    val value: ResponseAs[Choice[HttpError[String], DeserializationError[circe.Error], String], Nothing] =
+      asJson[String]
+    val value1 = value.map(unsafeHttp).merge2
+    val value2 = asJson[String].map(unsafeDeserialization).merge2
+    val value3 = asJson[String].map(unsafeHttp).map(unsafeDeserialization).merge2
```
Comment (ghostbuster91) [inline json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala:50]: `merge` method has to be called by the end user

File: json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala [hunk 4]
```diff
@@ -41,3 +68,58 @@ trait SttpCirceApi {
...
+  }
+  
+  implicit class RichResponseAs3[A,B,+S](v:ResponseAs[Choice[B,B,A],S])(implicit ev: B =:!= A){
+    def merge2: ResponseAs[Either[B,A], S] = {
+      v.map(s=> new PartiallyMergeableChoice2(s).merge)
+    }
+  }
+}
+
+sealed abstract class Choice[+L, +M, +R]() {
```
Comment (ghostbuster91) [inline json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala:110]: Should we change its name to something more domain specific like `Response` ?

**Comment by adamw** [comment]
What would the end-user API be like? Would be good to see some API examples of what's the goal of the changes

File: json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala (hunk 3 above)
Comment (ghostbuster91) [inline json/circe/src/main/scala/sttp/client/circe/SttpCirceApi.scala:50]: The example is here

**Pull request closed without merging**